lua/obsidian/cache.lua

```diff
@@ -0,0 +1,305 @@
+    end
+  end)
+end
+
+---Watches the vault for changes.
+---@param self obsidian.Cache
+local enable_filewatch = function(self)
+  local filewatch = require("obsidian.filewatch")
+  filewatch.watch(self.client.dir.filename, create_on_file_change_callback(self))
+
+  vim.api.nvim_create_autocmd({ "QuitPre", "ExitPre" }, {
+    callback = function()
+      filewatch.release_resources()
+    end,
+  })
+end
+
+---@param self obsidian.Cache
+local check_vault_cache = function(self)
+  check_file_exists(self.client.opts.cache.path, function(exists)
+    if exists then
+      vim.schedule(function()
+        check_cache_notes_are_fresh(self)
+      end)
+    else
+      vim.schedule(function()
+        self:index_vault()
+      end)
+    end
+  end)
+end
+
+---@param client obsidian.Client
+Cache.new = function(client)
+  local self = Cache.init()
+  self.client = client
+
+  if client.opts.cache.enable then
+    enable_filewatch(self)
+
+    check_vault_cache(self)
+  end
+
+  return self
+end
+
+--- Reads all notes in the vaults and returns the founded data.
+---@param client obsidian.Client
+---@return { [string]: obsidian.cache.CacheNote }
+local get_cache_notes_from_vault = function(client)
+  local interator = search.find(client.dir, "", nil)
+
+  ---@type { [string]: obsidian.cache.CacheNote }
+  local created_note_caches = {}
+
+  local notepath = interator()
+
+  --TODO add indexing progress
+  while notepath do
+    --TODO make async
+    local note = Note.from_file(notepath, { read_only_frontmatter = true })
+
+    local absolute_path = note.path.filename
+    local relative_path = absolute_path:gsub(client.dir.filename .. "/", "")
+
+    local file_stat = uv.fs_stat(absolute_path)
+    local last_updated
+
+    if type(file_stat) ~= "table" then
+      log.err(table.concat { "couldn't get file stat from file ", absolute_path })
+      last_updated = 0
+    else
+      last_updated = file_stat.mtime.sec
+    end
+
+    ---@type obsidian.cache.CacheNote
+    local note_cache = {
+      absolute_path = absolute_path,
+      aliases = note.aliases,
+      relative_path = relative_path,
+      last_updated = last_updated,
+    }
+
+    created_note_caches[relative_path] = note_cache
+
+    notepath = interator()
+  end
+
+  return created_note_caches
+end
+
+--- Reads all notes in the vaults and saves them to the cache file.
+---@param self obsidian.Cache
+Cache.index_vault = function(self)
+  if not self.client.opts.cache.enable then
+    log.error "The cache is disabled. Cannot index vault."
+  end
+
+  local founded_links = get_cache_notes_from_vault(self.client)
+
+  save_cache_notes_to_file(founded_links, self.client.opts.cache.path)
+
+  log.info "Vault was indexed succesfully."
+end
+
+---Reads the cache file from client.opts.cache.path and returns the loaded cache.
+---@param self obsidian.Cache
+---@return { [string]: obsidian.cache.CacheNote } Key is the relative path to the vault, value is the cache of the note.
```

Comment:
lua_ls says it's missing a return here, recommend you set it up, to catch things like this.
```suggestion
---@return { [string]: obsidian.cache.CacheNote }? Key is the relative path to the vault, value is the cache of the note.
```